crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,328 @@
+            index_additions.append(self.index.index_tx(added_tx));
+        }
+        for (&added_outpoint, added_txout) in &graph_additions.txout {
+            index_additions.append(self.index.index_txout(added_outpoint, added_txout));
+        }
+
+        IndexedAdditions {
+            graph_additions,
+            index_additions,
+        }
+    }
+
+    /// Insert a floating `txout` of given `outpoint`.
+    pub fn insert_txout(
+        &mut self,
+        outpoint: OutPoint,
+        txout: &TxOut,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let mut update = TxGraph::<A>::default();
+        let _ = update.insert_txout(outpoint, txout.clone());
+        self.apply_update(update)
+    }
+
+    /// Insert and index a transaction into the graph.
+    ///
+    /// `anchors` can be provided to anchor the transaction to various blocks. `seen_at` is a
+    /// unix timestamp of when the transaction is last seen.
+    pub fn insert_tx(
+        &mut self,
+        tx: &Transaction,
+        anchors: impl IntoIterator<Item = A>,
+        seen_at: Option<u64>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        let txid = tx.txid();
+
+        let mut update = TxGraph::<A>::default();
+        if self.graph.get_tx(txid).is_none() {
+            let _ = update.insert_tx(tx.clone());
+        }
+        for anchor in anchors.into_iter() {
+            let _ = update.insert_anchor(txid, anchor);
+        }
+        if let Some(seen_at) = seen_at {
+            let _ = update.insert_seen_at(txid, seen_at);
+        }
+
+        self.apply_update(update)
+    }
+
+    /// Insert relevant transactions from the given `txs` iterator.
+    ///
+    /// Relevancy is determined by the [`Indexer::is_tx_relevant`] implementation of `I`. Irrelevant
+    /// transactions in `txs` will be ignored.
+    ///
+    /// `anchors` can be provided to anchor the transactions to blocks. `seen_at` is a unix
+    /// timestamp of when the transactions are last seen.
+    pub fn insert_relevant_txs<'t>(
+        &mut self,
+        txs: impl IntoIterator<Item = &'t Transaction>,
+        anchors: impl IntoIterator<Item = A> + Clone,
+        seen_at: Option<u64>,
+    ) -> IndexedAdditions<A, I::Additions> {
+        // As mentioned by @LLFourn: This algorithm requires the transactions to be topologically
+        // sorted because most indexers cannot decide whether something is relevant unless you have
+        // first inserted its ancestors in the index. We can fix this if we instead do this:
+        // 1. insert all txs into the index. If they are irrelevant then that's fine it will just
+        //    not store anything about them.
+        // 2. decide whether to insert them into the graph depending on whether `is_tx_relevant`
+        //    returns true or not. (in a second loop).
+        let txs = txs
+            .into_iter()
+            .inspect(|tx| {
+                let _ = self.index.index_tx(tx);
```

Comment:
How do we sort the transactions if it's not guaranteed we have the position in block?

Reply:
I don't think the position in the block is necessary for sorting via ancestry. You just need to check the outpoints.

Reply:
Okay, I see what you mean. However I think your assumption about the indexer implementation is incorrect.

> Adding everything to the indexer seems like a hack, and can bloat up memory.

I think an indexer implementation should not save anything that is irrelevant to the index.

Reply:
> I think an indexer implementation should not save anything that is irrelevant to the index.

The current impl doesn't seem to be doing that. It's indexing all transactions thrown at it.

Reply:
Which implementation are you talking about? `SpkTxOutIndex` or `KeychainTxOutIndex`?

Reply:
> Which implementation are you talking about? SpkTxOutIndex or KeychainTxOutIndex?

Both are same. It all boils down to this macro here https://github.com/bitcoindevkit/bdk/blob/5e026cfd03143f497d8607c2b4ff0e3ec1757453/crates/chain/src/spk_txout_index.rs#L55-L68

As it says, it's adding all the Oupoints, txouts and spks for every transactions that is being indexed. Now imagine an wallet trying to find relevant transaction for it across 10,000 blocks.

Little napkin maths:
`OutPoint`: 36 bytes
`TxOut`: 41 bytes (assuming p2wsh)
`Script` : 33 bytes (assuming p2wsh).

So one index operation is adding  `(OP * 2 + Txout * 1 + Script * 2) = 179 bytes"  in the Indexer map. For 10k "irrelevant" txs indexed such a way is 1.79 MB of irrelevant data stored in memory. If you wanna scan the whole blockchain the number is linerly worse (currently at 826.58M transactions)

The doc comment says: 

`1. insert all txs into the index. If they are irrelevant, then that's fine it will just not store anything about them.` 

This is not factually correct. We are storing 180 bytes per output per transaction.

Reply:
Line 61. We are only storing txouts with spks that we are tracking in `spk_indicies`...